File: packages/agent-sdk/src/context.ts
/**
 * CapabilityContext: everything a handler may touch. Handlers are stateless
 * between invocations — all state arrives here or lives in platform stores
 * (agent-patterns.md).
 */

import type { Budget } from '@acp/protocol';
import type { Logger } from 'pino';
import type { ModelClient } from './model.js';
import type { Retriever } from './retriever.js';

export class CapabilityContext {
  readonly tenant: string;
  readonly taskId: string;
  readonly stepId: string;
  readonly capability: string;
  readonly delegatedToken: string | undefined;
  readonly budget: Budget | undefined;
  readonly model: ModelClient;
  readonly log: Logger;
  private readonly retriever: Retriever | undefined;

  constructor(fields: {
    tenant: string;
    taskId: string;
    stepId: string;
    capability: string;
    delegatedToken: string | undefined;
    budget: Budget | undefined;
    model: ModelClient;
    retriever: Retriever | undefined;
    log: Logger;
  }) {
    this.tenant = fields.tenant;
    this.taskId = fields.taskId;
    this.stepId = fields.stepId;
    this.capability = fields.capability;
    this.delegatedToken = fields.delegatedToken;
    this.budget = fields.budget;
    this.model = fields.model;
    this.retriever = fields.retriever;
    this.log = fields.log;
  }

  /** Citation-carrying retrieval under the step's delegated identity. `k` defaults to 8. */
  retrieve(query: string, options: { k?: number } = {}): Promise<Record<string, unknown>[]> {
    if (this.retriever === undefined) {
      return Promise.reject(
        new Error(
          "no retriever configured — pass one to Agent(...) or use the eval harness's" +
            ' fixture retriever in tests',
        ),
      );
    }
    if (this.delegatedToken === undefined) {
      return Promise.reject(
        new Error('step carries no delegated token — retrieval requires the delegated identity'),
      );
    }
    return this.retriever.search(this.delegatedToken, query, {
      k: options.k ?? 8,
      taskId: this.taskId,
      stepId: this.stepId,
    });
  }
}

File: python/acp-agent-sdk/src/acp_agent_sdk/context.py
"""CapabilityContext: everything a handler may touch. Handlers are
stateless between invocations — all state arrives here or lives in
platform stores (agent-patterns.md)."""

from dataclasses import dataclass
from typing import Any

import structlog

from acp_agent_sdk.model import ModelClient
from acp_agent_sdk.retriever import Retriever


@dataclass
class CapabilityContext:
    tenant: str
    task_id: str
    step_id: str
    capability: str
    delegated_token: str | None
    budget: dict[str, Any] | None
    model: ModelClient
    _retriever: Retriever | None
    log: structlog.stdlib.BoundLogger

    async def retrieve(self, query: str, *, k: int = 8) -> list[dict[str, Any]]:
        """Citation-carrying retrieval under the step's delegated identity."""
        if self._retriever is None:
            raise RuntimeError(
                "no retriever configured — pass one to Agent(...) or use the eval harness's"
                " fixture retriever in tests"
            )
        if self.delegated_token is None:
            raise RuntimeError(
                "step carries no delegated token — retrieval requires the delegated identity"
            )
        return await self._retriever.search(
            self.delegated_token, query, k=k, task_id=self.task_id, step_id=self.step_id
        )
